[pytorch] Minor: add GIL assert to PythonRpcHandler::handleExceptionGILHeld

jjlilley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33557 [pytorch] Minor: add GIL assert to PythonRpcHandler::handleExceptionGILHeld**

We should add GIL asserts in some places to keep assumptions documented.
This just adds one in an exception codepath as a placeholder for more.

This change also moves a #define from a .h to the .cpp to reduce scope.

Differential Revision: [D20005387](https://our.internmc.facebook.com/intern/diff/D20005387/)

Commit:
[pytorch] Minor: add GIL assert to PythonRpcHandler::handleExceptionGILHeld

We should add GIL asserts in some places to keep assumptions documented.
This just adds one in an exception codepath as a placeholder for more.

This change also moves a #define from a .h to the .cpp to reduce scope.

Differential Revision: [D20005387](https://our.internmc.facebook.com/intern/diff/D20005387/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -121,6 +137,7 @@ void PythonRpcHandler::handleException(const py::object& obj) {
 }
 
 void PythonRpcHandler::handleExceptionGILHeld(const py::object& obj) {
+  assert(PyGILState_Check()); // Ensure GIL is held
```
Use `TORCH_INTERNAL_ASSERT` ?

Commit:
Update on "[pytorch] Minor: add GIL assert to PythonRpcHandler::handleExceptionGILHeld"


We should add GIL asserts in some places to keep assumptions documented.
This just adds one in an exception codepath as a placeholder for more.

This change also moves a #define from a .h to the .cpp to reduce scope.

Differential Revision: [D20005387](https://our.internmc.facebook.com/intern/diff/D20005387/)

[ghstack-poisoned]

xush6528 commented on torch/csrc/distributed/rpc/python_rpc_handler.cpp:
```diff
@@ -1,13 +1,30 @@
+// A macro that grabs the GIL, profiling the acquisition time. The average GIL
+// acquisition time will be recorded in RpcAgent's getMetrics().
+#define PROFILE_GIL_SCOPED_ACQUIRE                                       \
+  std::chrono::time_point<std::chrono::high_resolution_clock> startTime; \
```
Lint message

jjlilley replied:
This change is just moving that existing # define from the .h to the .cpp, to reduce scope.

The lint complaint is about using # define in general, nothing specific that's addressable without somewhat bigger changes

pritamdamania87 replied:
Actually looks like a bug in the linter since we're not actually declaring a constant here but an entire codeblock.

Commit:
rebase on "[pytorch] Minor: add GIL assert to PythonRpcHandler::handleExceptionGILHeld"


We should add GIL asserts in some places to keep assumptions documented.
This just adds one in an exception codepath as a placeholder for more.

This change also moves a #define from a .h to the .cpp to reduce scope.

Differential Revision: [D20005387](https://our.internmc.facebook.com/intern/diff/D20005387/)

[ghstack-poisoned]
